File: tourism-ai-platform/ai-services/orchestrator-svc/app/tools/metrics.py
from __future__ import annotations

from typing import Dict

from prometheus_client import Histogram

_HISTOGRAMS: Dict[str, Histogram] = {}


def integration_histogram() -> Histogram:
    histogram = _HISTOGRAMS.get("integration_request_duration_seconds")
    if histogram is None:
        histogram = Histogram(
            "integration_request_duration_seconds",
            "Integration request duration in seconds",
            labelnames=("provider", "status"),
            buckets=(0.1, 0.3, 0.5, 1, 2, 5, 10),
        )
        _HISTOGRAMS["integration_request_duration_seconds"] = histogram
    return histogram

File: tourism-ai-platform/backend/src/modules/external/payments/metrics.ts
import { trace } from '@opentelemetry/api';
import { Histogram, register } from 'prom-client';

const HISTOGRAM_NAME = 'integration_request_duration_seconds';
const HISTOGRAM_HELP = 'Integration request duration in seconds';

const BUCKETS = [0.05, 0.1, 0.3, 0.5, 1, 2, 5, 10];

const existingHistogram = (): Histogram<string> | null => {
  try {
    return register.getSingleMetric(HISTOGRAM_NAME) as Histogram<string> | null;
  } catch {
    return null;
  }
};

export const integrationDurationMetric =
  existingHistogram() ??
  new Histogram({
    name: HISTOGRAM_NAME,
    help: HISTOGRAM_HELP,
    labelNames: ['provider', 'status'],
    buckets: BUCKETS,
  });

export const paymentsTracer = trace.getTracer('payments-integrations');
